pkg/server/tunnel.go

```diff
@@ -20,15 +20,55 @@ import (
 	"fmt"
 	"io"
 	"math/rand"
+	"net"
 	"net/http"
 	"sync"
 	"time"
 
 	"k8s.io/klog/v2"
+	commonmetrics "sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/common/metrics"
 	"sigs.k8s.io/apiserver-network-proxy/konnectivity-client/proto/client"
 	"sigs.k8s.io/apiserver-network-proxy/pkg/server/metrics"
 )
 
+type httpFrontend struct {
+	conn      net.Conn
+	closeOnce sync.Once
+	closed    chan struct{}
+}
+
+func (h *httpFrontend) Send(pkt *client.Packet) error {
+	const segment = commonmetrics.SegmentToClient
+	metrics.Metrics.ObservePacket(segment, pkt.Type)
+
+	if pkt.Type == client.PacketType_CLOSE_RSP {
+		h.close()
```

Comment:
Is there a reason we are dropping this potential error? I'm fine if we only return an error on the initial close and subsequent calls return null. (Eg https://go.dev/play/p/6C1_0QlYDRL). Ditto later places we are no longer returning errors.

Reply:
We never returned an error on close. The old `CloseHTTP` function just unconditionally returned `nil` (deleted code on old lines 89-93):

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/60b5f0e02800910cbb363668e285ec0e78d7cae1/pkg/server/tunnel.go#L89-L93